apollo-integration/src/test/java/com/apollographql/apollo/InterceptorTest.java

```diff
@@ -0,0 +1,432 @@
+package com.apollographql.apollo;
+
+
+import android.support.annotation.NonNull;
+
+import com.apollographql.android.impl.normalizer.EpisodeHeroName;
+import com.apollographql.android.impl.normalizer.type.Episode;
+import com.apollographql.apollo.api.Operation;
+import com.apollographql.apollo.api.Response;
+import com.apollographql.apollo.cache.normalized.Record;
+import com.apollographql.apollo.exception.ApolloException;
+import com.apollographql.apollo.exception.ApolloParseException;
+import com.apollographql.apollo.interceptor.ApolloInterceptor;
+import com.apollographql.apollo.interceptor.ApolloInterceptorChain;
+
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import javax.annotation.Nonnull;
+
+import okhttp3.MediaType;
+import okhttp3.OkHttpClient;
+import okhttp3.Protocol;
+import okhttp3.Request;
+import okhttp3.ResponseBody;
+import okhttp3.mockwebserver.MockResponse;
+import okhttp3.mockwebserver.MockWebServer;
+
+import static com.apollographql.apollo.interceptor.ApolloInterceptor.InterceptorResponse;
+import static com.google.common.truth.Truth.assertThat;
+
+public class InterceptorTest {
+
+  private ApolloClient client;
+  private MockWebServer mockWebServer;
+  private OkHttpClient okHttpClient;
+
+  private static final String FILE_EPISODE_HERO_NAME_WITH_ID = "EpisodeHeroNameResponseWithId.json";
+  private static final String FILE_EPISODE_HERO_NAME_CHANGE = "EpisodeHeroNameResponseNameChange.json";
+  private static final int TIMEOUT_SECONDS = 3;
+
+  @Before
+  public void setup() {
+    mockWebServer = new MockWebServer();
+    okHttpClient = new OkHttpClient.Builder().build();
+  }
+
+  @After
+  public void tearDown() {
+    try {
+      mockWebServer.shutdown();
+    } catch (IOException e) {
+
+    }
+  }
+
+  @Test
+  public void syncApplicationInterceptorCanShortCircuitResponses() throws IOException, ApolloException {
+    mockWebServer.shutdown();
```

Comment:
Yeah, we do need this, so as to ensure that response is indeed the intercepted response and not  from the server.